src/single_random_draw.rs

```diff
@@ -0,0 +1,166 @@
+//! This library provides efficient algorithms to compose a set of unspent transaction outputs
+//! (UTXOs).
+
+#[cfg(any(test, feature = "rand"))]
+use crate::errors::LibError;
+
+use bitcoin::Amount;
+use bitcoin::blockdata::transaction;
+use bitcoin::FeeRate;
+use bitcoin::TxOut;
+use bitcoin::Weight;
+use crate::CHANGE_LOWER;
+use rand::seq::SliceRandom;
+
+/// Randomly select coins for the given target by shuffling the utxo pool and
+/// taking UTXOs until the given target is reached.
+#[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    outputs: &mut Vec<TxOut>,
+    fee_rate: FeeRate,
+    rng: &mut R)
+-> Result<Option<Vec<TxOut>>, LibError> {
+    outputs.shuffle(rng);
+
+    let mut selected_effective_value: Amount = Amount::ZERO;
+    let mut i = 0;
+    let mut result = Vec::new();
+
+    // The required size of a change output must be at least
+    // larger than the target by CHANGE_LOWER.  Therefore,
+    // CHANGE_LOWER + target becomes the new lower bound for the
+    // selection.
+    while selected_effective_value < target + CHANGE_LOWER && i < outputs.len() {
+        let output: &TxOut = &outputs[i];
+        let value: Amount = output.value;
+
+        let outputs = outputs.iter().map(|txout| txout.script_pubkey.len());
+
+        let weight: Weight = transaction::predict_weight(
+            None::<Vec<transaction::InputWeightPrediction>>,
+            Some(outputs),
+        );
```

Comment:
Shouldn’t this calculate the input weight on basis of the specific `output`, because you could have outputs of different types?

Reply:
I updated this to calculate the specific weight of every output so far selected.  Is this what you had in mind?

Reply:
I meant that “effective_value” is a per-input property. It is calculated from the UTXO’s amount, the corresponding input’s weight, and the current feerate. It must be evaluated on a per-input basis because the effective value of individual UTXOs can turn negative if they are worth a small amount and the feerate is sufficiently high. At that point, you’d want to exclude them, because they don’t help you progress towards funding the transaction, but rather increase the total amount that needs to be provided by other inputs.

Reply:
> ..the corresponding input’s weight..

I'm not certain what is meant by the "corresponding input" here.  Is this the input that's to be constructed which consumes the UTXO?

Reply:
Yes

Reply:
Why is this not expressed in the bitcoin-core version?  AFAICT core only computes the weight of a group of UTXOs: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp#L220

Reply:
Perhaps I'm still misunderstanding.  I've updated the computation to be on a per-utxo basis similar to how core works.

Reply:
Bitcoin Core groups all UTXOs with the same scriptPubKey to form an OutputGroupt to opportunistically mitigate address reuse: if multiple payments have been received to the same address, they will generally be spent together instead of in multiple separate transactions that would then be revealed to have been sent by the same party.

Reply:
Cool, I'll take a look more into UTXO groups in core.  Maybe that's something that could be added to Rust Bitcoin?

Reply:
Fair enough.  I'm not understanding what is missing currently.  predict_weight() can take a set of inputs and outputs,  Currently I'm passing each utxo to predict_weight() and calculating the total.  Is this not correct?  I think predict_weight() should be used however to calculate the entire set of candidate utxos instead of individually adding to the total.

Reply:
s/calculating the total/adding to the total/g

Reply:
When you build a transaction, you must select sufficient funds for a) creating the recipient outputs, b)  fees for the transaction to have the intended queuing priority, and c) a sufficient budget for a change output.

You know in advance what amount the recipient outputs sum up to, and how much budget it takes to create a non-dust change output of a known type at a specific feerate—you can include these amounts in the selection target.

However, the fees for the inputs depend on the number and type of inputs that get selected. You _cannot_ include these in the selection target, because then your selection target _shifts_ every time you select or deselect an input, depending on the weight of that input and the current feerate. It’s really hard to hit a moving target.

This problem is mitigated by keeping track of the _effective values_ of the inputs you have selected rather than their full amount—by deducting the cost of spending the UTXO from the UTXO’s amount, you can account for the net amount that selecting the input moves you closer to the selection target, because _the input has “already paid for itself”_ when you select it. That’s the point of UTXO’s “effective value” and why it’s calculated from the specific UTXO’s corresponding input weight and the current feerate.

Reply:
I’m a bit at a loss here. It seems that I’m not finding the right words to explain the basic constraints and goals of Coin Selection to you. You may find it useful to read this [Stack Exchange answer](https://bitcoin.stackexchange.com/a/114847/5406) as well as the 3rd and 5th chapter of my [Master thesis](https://murch.one/erhardt2016coinselection.pdf).

Reply:
Thanks, I think I understand now.  From here the difficult part is to figure out how to construct the corresponding input for each utxo.  I suppose the simplest thing is just require three additional arguments to the selection algorithm ( [script_sig,](https://github.com/rust-bitcoin/rust-bitcoin/blob/1318ff88e525cd5ef3afd169c8dfd4934271e58c/bitcoin/src/blockdata/transaction.rs#L188) [sequence](https://github.com/rust-bitcoin/rust-bitcoin/blob/1318ff88e525cd5ef3afd169c8dfd4934271e58c/bitcoin/src/blockdata/transaction.rs#L193) and [witness](witness)) for which every corresponding input will be built with.  If my understanding is correct, there will be `n` inputs and two outputs (destination output, change output) and then the remainder of the fields to calculate the weight from.

Reply:
I've yet to re-write the tests.  However this is what I was thinking should work.  Basically construct the transaction, get the weight and ensure that the fee can be paid for by the change_output, if not keep adding inputs to the transaction.

```
    utxos.shuffle(rng);

    let zipped = utxos.iter().zip(inputs.iter());
    let mut inputs:Vec<TxIn> = Vec::new();
    let mut value = Amount::ZERO;

    for (u, i) in zipped {
        value += u.value;
        inputs.push(i.clone());

        if value > target + CHANGE_LOWER {

            // TODO use chedcked_sub
            let change_amount = value - target - CHANGE_LOWER;

            let change_output = TxOut {
                value: change_amount,
                script_pubkey: script_pubkey.clone()
            };

            let destination_output = TxOut {
                value: target,
                script_pubkey: script_pubkey.clone()
            };
            
            transaction.input = inputs.clone();
            transaction.output = vec![change_output, destination_output];
                
            let weight = transaction.weight();
            let transaction_fee: Option<Amount> = fee_rate.checked_mul_by_weight(weight);
    
            match transaction_fee {
                Some(f) => {
                        
                    // there is not enough to pay the fee out of the change output
                    // without creating a dust output.  Remove the last input so
                    // another permutation can be tried.
                    if change_amount < f - CHANGE_LOWER{
                        inputs.pop();
                        transaction.input = inputs.clone();
                        value -= u.value;
                    } else {
                        return Ok(Some(transaction.clone()))
                    }

                },
                None => return Err(Error::Multiplication(weight, fee_rate)),
            }
        }
    }

    Ok(None)
}
```

Reply:
As far as I can tell, this will skip UTXOs with negative effective value, but still accounts for inputs with the whole value rather than the effective value towards the target. If my understanding of your code is right, 

        value += u.value;
        
should read
        
        value += u.value - input_fee;

I suggest that you implement some tests that check whether your code selects sufficient funds to both pay the fee and still exceed `target + CHANGE_LOWER` with its selection.

Reply:
Thanks!  I've updated the tests and the logic as suggested.  Also I've also added an assertion to test that the change amount is greater than the dust limit once fees are subtracted [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23/files#diff-ce2c6b6b3132bf7b26a80fc104adfdeb0eb023b9bfd487c0da95d83830a73a32R146)

Reply:
I believe this has been resolved now.